diff -Naur irc2.10.3p1/common/os.h irc2.10.3p1-p/common/os.h
--- irc2.10.3p1/common/os.h	Wed Jun  7 00:32:49 2000
+++ irc2.10.3p1-p/common/os.h	Thu Aug 24 16:27:35 2000
@@ -137,6 +137,9 @@
 /* Linux 2.1.xx supports poll(), header files are not upto date yet */
 #  define POLLRDNORM 0x0040
 # endif
+# if linux && !defined(POLLWRNORM)
+#  define POLLWRNORM 0x0100
+# endif
 #endif
 
 #if HAVE_STROPTS_H
diff -Naur irc2.10.3p1/support/Makefile.in irc2.10.3p1-p/support/Makefile.in
--- irc2.10.3p1/support/Makefile.in	Wed Jun  7 00:35:16 2000
+++ irc2.10.3p1-p/support/Makefile.in	Thu Aug 24 16:27:35 2000
@@ -67,21 +67,21 @@
 # Directories definitions
 #
 # Directory in which to install irc.
-client_bin_dir = @bindir@
+client_bin_dir = $(DESTDIR)@bindir@
 # Directory in which to install ircd, iauth, ircdwatch, mkpasswd and chkconf.
-server_bin_dir = @sbindir@
+server_bin_dir = $(DESTDIR)@sbindir@
 # Directory in which to install the client manual page.
-client_man_dir = @mandir@/man1
+client_man_dir = $(DESTDIR)@mandir@/man1
 # Directory in which to install the configuration manual page.
-conf_man_dir = @mandir@/man5
+conf_man_dir = $(DESTDIR)@mandir@/man5
 # Directory in which to install the server manual page.
-server_man_dir = @mandir@/man8
+server_man_dir = $(DESTDIR)@mandir@/man8
 # Directory where config files (ircd.conf, ircd.motd and iauth.conf) live.
-ircd_conf_dir = @sysconfdir@
+ircd_conf_dir = $(DESTDIR)@sysconfdir@
 # Directory where state files (ircd.pid, ircd.tune) live.
-ircd_var_dir = @localstatedir@
+ircd_var_dir = $(DESTDIR)@localstatedir@
 # Directory where log files (users, opers, rejects and auth) live.
-ircd_log_dir = @logdir@
+ircd_log_dir = $(DESTDIR)@logdir@
 
 #
 # Most of these PATHs are hardcoded in the binaries.
