diff -ur irc2.11.0b15.orig/support/config.h.dist irc2.11.0b15/support/config.h.dist
--- irc2.11.0b15.orig/support/config.h.dist	2004-10-27 14:51:38.000000000 +0200
+++ irc2.11.0b15/support/config.h.dist	2004-10-27 14:52:19.774516864 +0200
@@ -35,7 +35,7 @@
  * may want to log when ircdwatch reloads the config or when
  * ircd croaks and ircdwatch has to restart it
  */
-#define IRCDWATCH_USE_SYSLOG
+#undef IRCDWATCH_USE_SYSLOG
 #ifdef IRCDWATCH_USE_SYSLOG
 # define IRCDWATCH_SYSLOG_IDENT     "ircdwatch"
 # define IRCDWATCH_SYSLOG_OPTIONS   (LOG_PID)
@@ -106,7 +106,7 @@
  * Define this if you want to have TKLINE and UNTKLINE commands
  * for opers and services.
  */
-#define TKLINE
+#undef TKLINE
 
 /* 
  * Operator rights can be precisely defined in O:line flags. However
@@ -133,7 +133,7 @@
  * 1 server = 1 connection, 1 user = 1 connection.
  * Due to various sanity checks during startup, minimum is 13.
  */
-#define MAXCONNECTIONS	50
+#define MAXCONNECTIONS	512
 
 /* MAXIMUM LINKS
  *
@@ -170,7 +170,7 @@
  * A HUB Server has many servers connect to it at the same as opposed
  * to a leaf which just has 1 server (typically the uplink).
  */
-#undef	HUB
+#define	HUB
 
 #ifdef HUB
 /*
@@ -257,12 +257,12 @@
 /*
  * Define this if you want to log failed /oper attempts.
  */
-#undef FAILED_OPERLOG
+#define FAILED_OPERLOG
 
 /*
  * Define this to make ircd create logfiles if they do not exist.
  */
-#undef LOGFILES_ALWAYS_CREATE
+#define LOGFILES_ALWAYS_CREATE
 
 /*
  * Define this if you want to use crypted passwords for operators in your
@@ -307,8 +307,8 @@
  * This should only be defined if you are running as root...
  * and even then perhaps not.
  */
-/* #undef	IRC_UID	65534 */
-/* #undef	IRC_GID	65534 */
+#define	IRC_UID	75
+#define	IRC_GID	75
 
 /*
  * CLIENT_FLOOD
@@ -352,7 +352,7 @@
  * The library and the include files must have been found by configure,
  * if you have installed the zlib after running configure, run it again.
  */
-#undef ZIP_LINKS
+#define ZIP_LINKS
 
 /*
  * Defining this will add an artificial 2 seconds delay for accepting
@@ -473,8 +473,7 @@
 /* Following notice is sent before and after /LIST output.
  * If you do not want such behaviour, undefine.
  */
-#define LIST_ALIS_NOTE "Usage of /list for listing all channels is " \
-		       "deprecated. Please use \"/squery alis help\" instead."
+/* #undef LIST_ALIS_NOTE "Usage of /list for listing all channels is deprecated. Please use \"/squery alis help\" instead." */
 
 /* Define this to see when a channel topic was set and who it was set by.
 ** Warning: eats memory (around 5MB)
@@ -498,8 +497,8 @@
  *
  * Defining to 0 disables entering split-mode.
  */
-#define SPLIT_USERS	70000
-#define SPLIT_SERVERS	30
+#define SPLIT_USERS	0
+#define SPLIT_SERVERS	0
 
 /*
 ** Notice sent to connecting users if the server is in the split-mode.
