diff -ur irc2.10.3p7.orig/ircd/ircd.c irc2.10.3p7/ircd/ircd.c
--- irc2.10.3p7.orig/ircd/ircd.c	2004-05-09 22:09:08.000000000 +0200
+++ irc2.10.3p7/ircd/ircd.c	2004-10-11 21:35:39.372299576 +0200
@@ -221,6 +221,7 @@
 	aConfItem *con_conf = NULL;
 	double	f, f2;
 	aCPing	*cp;
+	int allheld = 1;
 
 	Debug((DEBUG_NOTICE,"Connection check at   : %s",
 		myctime(currenttime)));
@@ -246,6 +247,7 @@
 		if (aconf->port <= 0)
 			continue;
 
+		allheld = 0;
 		cltmp = Class(aconf);
 		confrq = get_con_freq(cltmp);
 		aconf->hold = currenttime + confrq;
@@ -292,6 +294,7 @@
 				    con_conf->name, con_conf->host);
 	    }
 	else
+	if (allheld == 0)
 	{
 		/* No suitable conf for AC was found, so why bother checking
 		** again? If some server quits, it'd get reenabled --B. */
