diff -ur irc2.11.0b14.orig/common/os.h irc2.11.0b14/common/os.h
--- irc2.11.0b14.orig/common/os.h	2004-10-02 03:16:16.000000000 +0200
+++ irc2.11.0b14/common/os.h	2004-10-11 19:34:34.092789944 +0200
@@ -139,6 +139,9 @@
 /* Linux 2.1.xx supports poll(), header files are not upto date yet */
 #  define POLLRDNORM 0x0040
 # endif
+# if linux && !defined(POLLWRNORM)
+#  define POLLWRNORM 0x0100
+# endif
 #endif
 
 #if HAVE_STROPTS_H
diff -ur irc2.11.0b14.orig/support/Makefile.in irc2.11.0b14/support/Makefile.in
--- irc2.11.0b14.orig/support/Makefile.in	2004-10-04 21:55:34.000000000 +0200
+++ irc2.11.0b14/support/Makefile.in	2004-10-11 19:35:50.530169696 +0200
@@ -69,17 +69,17 @@
 # Directories definitions
 #
 # Directory in which to install ircd, iauth, ircdwatch, mkpasswd and chkconf.
-server_bin_dir = @sbindir@
+server_bin_dir = $(DESTDIR)@sbindir@
 # Directory in which to install the configuration manual page.
-conf_man_dir = @mandir@/man5
+conf_man_dir = $(DESTDIR)@mandir@/man5
 # Directory in which to install the server manual page.
-server_man_dir = @mandir@/man8
+server_man_dir = $(DESTDIR)@mandir@/man8
 # Directory where config files (ircd.conf, ircd.motd and iauth.conf) live.
-ircd_conf_dir = @sysconfdir@
+ircd_conf_dir = $(DESTDIR)@sysconfdir@
 # Directory where state files (ircd.pid, ircd.tune) live.
-ircd_var_dir = @rundir@
+ircd_var_dir = $(DESTDIR)@rundir@
 # Directory where log files (users, opers, rejects and auth) live.
-ircd_log_dir = @logdir@
+ircd_log_dir = $(DESTDIR)@logdir@
 
 #
 # Most of these PATHs are hardcoded in the binaries.
