diff -ur irc2.11.0b14.orig/ircd/s_user.c irc2.11.0b14/ircd/s_user.c
--- irc2.11.0b14.orig/ircd/s_user.c	2004-10-07 02:16:47.000000000 +0200
+++ irc2.11.0b14/ircd/s_user.c	2004-10-11 23:46:31.655571072 +0200
@@ -2877,15 +2877,21 @@
 int	m_ping(aClient *cptr, aClient *sptr, int parc, char *parv[])
 {
 	aClient *acptr;
-	char	*destination;
+	char	*origin, *destination;
 
+	origin = parv[1];
 	destination = parv[2]; /* Will get NULL or pointer (parc >= 2!!) */
 
+	acptr = find_client(origin, NULL);
+	if (!acptr)
+		acptr = find_server(origin, NULL);
+	if (!acptr || acptr != sptr)
+		origin = cptr->name;
 	if (!BadPtr(destination) && match(destination, ME) != 0)
 	{
 		if ((acptr = find_server(destination, NULL)))
 			sendto_one(acptr, ":%s PING %s :%s", parv[0],
-				cptr->name, destination);
+				origin, destination);
 	    	else
 		{
 			sendto_one(sptr, replies[ERR_NOSUCHSERVER],
