diff -ur irc2.11.0b14.orig/support/config.h.dist irc2.11.0b14/support/config.h.dist
--- irc2.11.0b14.orig/support/config.h.dist	2004-10-06 14:20:12.000000000 +0200
+++ irc2.11.0b14/support/config.h.dist	2004-10-11 19:42:06.952944736 +0200
@@ -457,11 +457,7 @@
 /*
 ** Define delimiter of fields in ircd.conf.
 */
-#ifdef INET6
-# define IRCDCONF_DELIMITER '%'
-#else
-# define IRCDCONF_DELIMITER ':'
-#endif
+#define IRCDCONF_DELIMITER '%'
 
 /*
  * Max number of channels a user is allowed to join.
