diff -urN irc2.10.3.orig/support/config.h.dist irc2.10.3/support/config.h.dist
--- irc2.10.3.orig/support/config.h.dist	Tue Sep  7 21:37:13 1999
+++ irc2.10.3/support/config.h.dist	Wed Sep  8 14:10:38 1999
@@ -118,7 +118,7 @@
  * if 'OPER_x' is undefined so you can't give local operators more rights
  * than global ones.
  */
-#undef	OPER_KILL
+#define	OPER_KILL
 #define	OPER_REHASH
 #undef	OPER_RESTART
 #define	OPER_DIE
@@ -299,8 +299,8 @@
  * define IRC_UID to that UID.  This should only be defined if you are running
  * as root and even then perhaps not.
  */
-#undef	IRC_UID
-#undef	IRC_GID
+#define	IRC_UID 75
+#define	IRC_GID 75
 
 #ifdef	notdef
 #define	IRC_UID	65534	/* eg for what to do to enable this feature */
@@ -352,7 +352,7 @@
  * The library and the include files must have been found by configure,
  * if you have installed the zlib after running configure, run it again.
  */
-#undef ZIP_LINKS
+#define ZIP_LINKS
 
 /*
  * Defining this will add an artificial 2 seconds delay for accepting
diff -urN irc2.10.3.orig/support/iauth.conf irc2.10.3/support/iauth.conf
--- irc2.10.3.orig/support/iauth.conf	Tue Sep  7 21:37:13 1999
+++ irc2.10.3/support/iauth.conf	Wed Sep  8 14:11:12 1999
@@ -5,7 +5,8 @@
 #
 
 # If iauth timeouts, then reject user
-notimeout
+# notimeout
+
 # This makes the IRC server require that iauth performs the authentication
 # in order for a new user connection to be accepted
 required
