diff -ur irc2.10.3p6/support/config.h.dist irc2.10.3p6-conf/support/config.h.dist
--- irc2.10.3p6/support/config.h.dist	2004-04-29 18:33:11.797559848 +0200
+++ irc2.10.3p6-conf/support/config.h.dist	2004-04-29 18:34:05.033466760 +0200
@@ -35,7 +35,7 @@
  * may want to log when ircdwatch reloads the config or when
  * ircd croaks and ircdwatch has to restart it
  */
-#define IRCDWATCH_USE_SYSLOG
+#undef IRCDWATCH_USE_SYSLOG
 #ifdef IRCDWATCH_USE_SYSLOG
 # define IRCDWATCH_SYSLOG_IDENT     "ircdwatch"
 # define IRCDWATCH_SYSLOG_OPTIONS   (LOG_PID)
@@ -120,7 +120,7 @@
  */
 #define	OPER_KILL
 #define	OPER_REHASH
-#undef	OPER_RESTART
+#define	OPER_RESTART
 #undef	OPER_DIE
 #define	OPER_REMOTE
 #undef	LOCOP_REHASH
@@ -138,7 +138,7 @@
  * This should be at *least* 4: 2 listen ports (1 tcp, 1 udp)
  *				1 dns port, 1 client
  */
-#define MAXCONNECTIONS	16000
+#define MAXCONNECTIONS	512
 
 /* MAXIMUM LINKS
  *
@@ -175,7 +175,7 @@
  * A HUB Server has many servers connect to it at the same as opposed
  * to a leaf which just has 1 server (typically the uplink).
  */
-/* #define	HUB */
+#define	HUB
 
 #ifdef HUB
 /*
@@ -275,8 +275,8 @@
  * define IRC_UID to that UID.  This should only be defined if you are running
  * as root and even then perhaps not.
  */
-#undef	IRC_UID
-#undef	IRC_GID
+#define	IRC_UID	75
+#define	IRC_GID	75
 
 #ifdef	notdef
 #define	IRC_UID	65534	/* eg for what to do to enable this feature */
@@ -458,7 +458,7 @@
    use ALIS instead, define this. Notice is sent BEFORE any output.
  */
  
-#define LIST_ALIS_NOTE "Usage of /list for listing all channels is deprecated. Please use \"/squery alis help\" instead."
+#undef LIST_ALIS_NOTE
 
 /* If you want to send server IS_SUPPORT upon connect and along with
    VERSION reply. Also moves redirect numerics from 5 to 10.
@@ -489,8 +489,8 @@
 #define SPLIT_HANDLE
 
 /* defines minimum for GLOBAL channels */
-#define SPLIT_SERV  20
-#define SPLIT_USERS 40000
+#define SPLIT_SERV  0
+#define SPLIT_USERS 0
 
 /* Delay binding to listening ports with 'D' in second field until server 
  * is out of split.
@@ -519,7 +519,7 @@
 
 /* define file which is used for saving k-lines added by /kline
  */
-#undef KLINE_FILE IRCDCONF_PATH
+//#undef KLINE_FILE IRCDCONF_PATH
 
 /* undef if you don't want to have Denied connection  because of open proxy */
 #undef IAUTH_VERBOSE_REJECTS
@@ -539,7 +539,7 @@
 ** You may want to define this to disable "STATS k" for plain users
 ** and redirect them to some WWW page.
 */
-#define NOSTATSK "Please use http://www.irc.pl/kline instead."
+#undef NOSTATSK
 
 /*   STOP STOP STOP STOP STOP STOP STOP STOP STOP STOP STOP STOP STOP STOP  */
 /*   STOP STOP STOP STOP STOP STOP STOP STOP STOP STOP STOP STOP STOP STOP  */
