diff -ur irc2.10.3p7.orig/support/config.h.dist irc2.10.3p7/support/config.h.dist
--- irc2.10.3p7.orig/support/config.h.dist	2004-10-11 22:26:58.000000000 +0200
+++ irc2.10.3p7/support/config.h.dist	2004-10-11 22:29:52.853695224 +0200
@@ -238,7 +238,7 @@
  * define this if you want to use crypted passwords for operators in your
  * ircd.conf file. See contrib/mkpasswd/README for more details on this.
  */
-#undef	CRYPT_OPER_PASSWORD
+#define	CRYPT_OPER_PASSWORD
 
 /*
  * If you want to store encrypted passwords in N-lines for server links,
